Global Styles: Try style system at site edit

oandregal opened the pull request:
This is a fork and supersedes https://github.com/WordPress/gutenberg/pull/20061 by @itsjonq It adds a Global Styles sidebar with typography controls in the Site Editor.

![2020-03-30-global-styles-edit-site](https://user-images.githubusercontent.com/583546/77948908-6ca40100-72c6-11ea-8f73-1d7b4a6825f7.gif)

A few notes:

- The is not the final UI, we'll iterate in further PRs until we get to https://github.com/WordPress/gutenberg/issues/21030
- It focuses on the global typography first, leaving other things for subsequent PRs.
- It adds the mechanics to read/update user & theme data via the Entities API.
- It removes the use of the `wp-gs` class in favor of CSS variables in the blocks. Following suit on what was done for colors https://github.com/WordPress/gutenberg/pull/21021 and line-height https://github.com/WordPress/gutenberg/pull/20775 this PR adds the font-size & font-weight CSS variables to Paragraph and Heading blocks.
- It loads the CSS variables in both edit-site & edit-post ([see](https://github.com/WordPress/gutenberg/pull/20530#discussion_r400827213)).
- It leaves out the variable generation mechanism / variable calculation to further explore in a subsequent PR. See [comment](https://github.com/WordPress/gutenberg/pull/20530#discussion_r391767806).
- It uses the new interface package to add the global styles sidebar.

## Test instructions

Test in `edit-site`:

- Install and activate the [Global Styles demo theme](https://github.com/WordPress/theme-experiments/pull/22).
- Enable the FSE Experiment under "Gutenberg > Experiments".
- Go to "Gutenberg > Site Editor (beta)" and play with the Global Styles controls in the sidebar.
- Save the changes to the Global Styles template via the "Update design" button in the editor.
- Go to the front and see that your changes are applied and are the same as the editor.

Test in `edit-post`:

- Install and activate the [Global Styles demo theme](https://github.com/WordPress/theme-experiments/pull/22).
- Enable the FSE Experiment under "Gutenberg > Experiments".
- Go to the post editor and create some content (paragraphs & headings). Make sure the values saved in the previous test are honored (inspect in the browser devtools, perhaps).
- Edit the line-height of a paragraph or heading, and make sure it is updated.

Commit:
Create initial GlobalStyle component within SiteEdit

oandregal committed:
Iterate on Global Styles for site-edit

nosolosw commented:
I'm open to create a separate PR for the color-controls, if that helps ship this faster. Happy to merge it as part of this as well.

ItsJonQ commented:
@nosolosw Amazing work on this so far! Thank you for taking what I started and extending it! I'm comfortable closing my PR to continue with this one.

---

After poking around a bit, these are some of the things I noticed (not your fault! Just observations)

## Infrastructure
##### ⭐️Need core values for `typography`

The lack of these values is causing themes with an empty `experimental-theme.json` to break.

##### ⭐️`experimental-theme.json` from a theme directory is **REPLACING** core style values. Needs to **EXTEND**.

Values that are provided by a theme's `experimental-theme.json` are completely erasing the values provided by core.

## Rendering

##### ⭐️Remove computed values (line-height, font-sizes) for now (?) (in the JS logic)

Ideally, I hope that we're able to generated computed values from a single value. For example, `fontSize` -> all heading sizes. At the moment, I did this in the JS layer during updates. I feel like for this mechanism to be successful, it may have to be done server side (as well... maybe?).

It works okay *if* the computed values hook/function is able to execute - either on change, or maybe even on initial Edit Site load. However, if that doesn't happen, then no computed values happen.

##### ⭐️Styles not being rendered (Gutenberg or Frontend) by default. Currently only works if specified by the theme.

No style updates occur if we switch to theme without the CSS variables hooked it, example. TwentyTwenty (assuming `experimental-theme.json`). Core blocks should be hooked up to these variables assuming global styles are activated with dedicated global styles CSS.

## Theme: Styles
#####  ⭐️Lack of `.wp-gs` env scope is causing render issues in Gutenberg UI

An example of this is would be paragraphs when activating the (experimental) Global Styles theme:

<img width="920" alt="Screen Shot 2020-02-28 at 1 45 58 PM" src="https://user-images.githubusercontent.com/2322354/75577780-b764fb80-5a30-11ea-91aa-e799cdf794e6.png">

nosolosw commented:
Nice feedback @itsjonq ! I've got all ready for a second round. To address each point separately:

- Need core values for typography. Added. See `experimental-default-global-styles.json`.
- Theme needs to extend core values. Fixed.
- Computed values. I've added the same generation logic server-side for the moment. I'd like to explore this fully in a separate PR, including the possibility of perhaps creating a new endpoint to centralize all the generation logic in the server so mobile can reuse as well.
- Theme needs to specify block styles with CSS variables. This is intentional so we can iterate faster by not messing with core styles at the moment. Note, however, that the [demo playground theme](https://github.com/WordPress/theme-experiments/pull/22/files) that comes with this has separate stylesheets per block, so we can merge them back in core easily.
- Lack of `.wp-gs` env scope. While the styles are external, we don't need this. However, what you spotted is true! The reason is that, unlike `edit-post`, `edit-site` doesn't wrap the styles using the `editor-styles-wrapper`. We need to fix that --- was thinking a different PR would be best.

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -164,15 +164,90 @@ function gutenberg_experimental_global_styles_get_theme() {
+		$global_styles,
+		array(
+			'typography' => array(
+				'line-height-heading' => gutenberg_experimental_global_styles_generate_line_height( $line_height, 0.8 ),
```
Can't we use CSS calc https://developer.mozilla.org/en-US/docs/Web/CSS/calc on the CSS end to generate these values?

E.g: core would set variable for font-size-heading-3 equal to calc depending on another variable. Themes would be able to overwrite that.

nosolosw replied:
See https://github.com/WordPress/gutenberg/issues/21030

nosolosw replied:
See https://github.com/WordPress/gutenberg/pull/20530#discussion_r391767806

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -164,15 +164,90 @@ function gutenberg_experimental_global_styles_get_theme() {
+				'line-height-heading' => gutenberg_experimental_global_styles_generate_line_height( $line_height, 0.8 ),
+				'font-size-heading-1' => gutenberg_experimental_global_styles_generate_font_size( $font_size, $font_scale, 5 ),
+				'font-size-heading-2' => gutenberg_experimental_global_styles_generate_font_size( $font_size, $font_scale, 4 ),
+				'font-size-heading-3' => gutenberg_experimental_global_styles_generate_font_size( $font_size, $font_scale, 3 ),
```
Would we have a UI to change all these values, or the user will not be able to change them?

nosolosw replied:
The current UI for font-size is this https://github.com/WordPress/gutenberg/issues/21030 It only contains base and scale, so the other sizes should adapt from there.

I was also a bit concerned about global styles being opinionated about a style computation (which is styles/theme territory in my view). I'll give it a shot at making global styles agnostic about style computation, and pushing those to the edges (CSS land).

nosolosw replied:
As a first step, I've pushed [computations](https://github.com/WordPress/gutenberg/pull/20530/files?file-filters%5B%5D=.scss#diff-f2f5efb13e0e16e61312aa9b0c028ae1R26) based on font base & font scale to CSS land. I believe we want to evolve the theme.json to be able to do more things, but I'd like to do that in a subsequent PR --- otherwise, this will get too big to review.

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -164,15 +164,90 @@ function gutenberg_experimental_global_styles_get_theme() {
+		array(
+			'typography' => array(
+				'line-height-heading' => gutenberg_experimental_global_styles_generate_line_height( $line_height, 0.8 ),
+				'font-size-heading-1' => gutenberg_experimental_global_styles_generate_font_size( $font_size, $font_scale, 5 ),
```
If the theme explicitly sets "font-size-heading-1" aren't we overwriting the value with our automatic computation?

nosolosw replied:
See https://github.com/WordPress/gutenberg/issues/21030

nosolosw replied:
See https://github.com/WordPress/gutenberg/pull/20530#discussion_r391767806

jorgefilipecosta commented on packages/global-styles/src/provider.js:
```diff
@@ -0,0 +1,189 @@
+
+// Make object reference immutable
+// so it stays the same in the function setters.
+let styles = {};
```
Should we move this inside the hook? And use the hook useRef with a start value of {}?

ZebulanStanphill replied:
Also, the comment says "Make object reference immutable", but it's using `let`, not `const`.

jorgefilipecosta replied:
cc: @nosolosw should we do some update here?

jorgefilipecosta commented on packages/global-styles/src/provider.js:
```diff
@@ -0,0 +1,189 @@
+	};
+}
+
+function generateFontSizesHeading( { fontSize, fontScale } ) {
```
We are essentially duplicating the logic in JavaScript and PHP, if we could just use CSS calc we would simplify the things.

nosolosw replied:
See https://github.com/WordPress/gutenberg/issues/21030

nosolosw replied:
See https://github.com/WordPress/gutenberg/pull/20530#discussion_r391767806

jorgefilipecosta commented on packages/global-styles/src/renderer.js:
```diff
@@ -0,0 +1,82 @@
+	}, [ generatedStyles ] );
+}
+
+function useGlobalStylesEnvironment() {
```
Would it be possible to add the wp-gs class directly on the server if global styles are supported?

nosolosw replied:
We're no longer relying on the `wp-gs` so I've removed the code that dealt with that.

jorgefilipecosta commented on packages/global-styles/src/provider.js:
```diff
@@ -0,0 +1,189 @@
+	// so controls can modify the styles.
+	const { editEntityRecord } = useDispatch( 'core' );
+
+	const setColor = ( newStyles ) => {
```
We are generating new setters on every render. Should we consider the usage of useCallback/useMemo?

jorgefilipecosta reviewed (commented):
Hi @nosolosw thank you for the work on this PR 👍 
The work on global styles is complex with multiple moving parts, so this PR is not simple to review.
I wonder if we could simplify things a big by separating the PR into multiple PR's.

- A PR with the new color components. That's a totally separate work where we will need to deal with compatibility with current usages, the color palette, etc.

- A PR with the basis for global styles but dealing only with a variable for example line-height.

- A PR dealing with some additional variables and maybe the way to compute a variable from another.

What do you think, would this separation make sense?

nosolosw commented on packages/components/src/color-control/index.js:
```diff
@@ -0,0 +1,106 @@
+/**
```
Agreed. I've removed the color bits from this PR and will let @ItsJonQ (as the original author) to spin up a new PR the color component.

johnstonphilip commented:
Just confirming, does this PR only handle typography and not colors at the moment? It also does not appear to pull from my experimental-theme.json at the moment. 

<img width="286" alt="Screen Shot 2020-03-26 at 4 42 36 PM" src="https://user-images.githubusercontent.com/7538525/77694741-f4190980-6f80-11ea-8c21-9f1ff5bd0d74.png">

nosolosw commented:
> Just confirming, does this PR only handle typography and not colors at the moment? It also does not appear to pull from my experimental-theme.json at the moment.

Yeah, I've reduced scope based on feedback. This is now in a transitioning state while I apply other improvements, will report when it's ready for review/test.

oandregal committed:
Make the typography panel open by default

oandregal committed:
Load assets for wp-admin including the edit-post

nosolosw commented on lib/global-styles.php:
```diff
@@ -148,7 +148,7 @@ function gutenberg_experimental_global_styles_get_user_cpt_id() {
  */
 function gutenberg_experimental_global_styles_get_core() {
 	return gutenberg_experimental_global_styles_get_from_file(
-		dirname( dirname( __FILE__ ) ) . '/experimental-default-global-styles.json'
```
This was moved from the top-level to `lib/global-styles` as per [this comment](https://github.com/WordPress/gutenberg/pull/19883#discussion_r392226550).

nosolosw commented on lib/global-styles.php:
```diff
@@ -189,16 +195,6 @@ function gutenberg_experimental_global_styles_resolver( $global_styles ) {
 	return $css_rule;
 }
 
-/**
```
Removing this check means that, if the FSE experiment is enabled, we'll load the CSS variables in all wp-admin, not only in edit-site pages.

I did this because the CSS vars are now inlined into the block's CSS, so if I load the edit-post I'd expect it to work the same as the edit-site (variables are loaded, so blocks look the same).

Another option would be to only load this in the edit-post & edit-site? I don't think that's necessary but happy to hear other thoughts. cc @jorgefilipecosta

ZebulanStanphill replied:
It's worth noting that if #21102 is merged, the editor content will be iframed, so CSS variables outside of that iframe would have no effect.

jasmussen replied:
>  so CSS variables outside of that iframe would have no effect.

Something to be mindful of in the iframe PR. We have ways to communicate into and out of the iframe, so it should be possible to set things properly in both.

mtias commented on packages/global-styles/src/controls/typography-controls.js:
```diff
@@ -0,0 +1,80 @@
+				step={ 0.025 }
+				onChange={ ( value ) => setTypography( { fontScale: value } ) }
+			/>
+			<h2>Heading</h2>
```
Let's translate these.

nosolosw replied:
done at https://github.com/WordPress/gutenberg/pull/20530/commits/fc5da4766acf77fa4b49d3f0419416d1bb58dc66

mtias commented on packages/edit-site/src/components/sidebar/index.js:
```diff
@@ -25,11 +26,11 @@ function Sidebar() {
+				icon={ typography }
 			>
-				<p>Global Styles area</p>
+				<GlobalStylesControls bubblesVirtually />
```
Let's render this one before the "settings" tool.

nosolosw replied:
done at https://github.com/WordPress/gutenberg/pull/20530/commits/f1ba4d88892b37b18483f99ef29be745bff3fde2

nosolosw commented on lib/global-styles.php:
```diff
@@ -224,37 +214,6 @@ function gutenberg_experimental_global_styles_enqueue_assets() {
 	wp_enqueue_style( 'global-styles' );
 }
 
-/**
```
We're no longer relying on the `wp-gs` class.

nosolosw commented on lib/global-styles/experimental-default-global-styles.json:
```diff
@@ -0,0 +1,14 @@
+{
```
Moved from the top-level. So posting here the old contents for comparison:

```
{
	"color": {
		"primary": "#52accc",
		"background": "white",
		"text": "black"
	}
}
```

nosolosw replied:
Note that I've removed `primary` here as the color hooks at https://github.com/WordPress/gutenberg/pull/21021 don't implement it.

nosolosw commented on packages/block-library/src/button/style.scss:
```diff
@@ -36,10 +36,6 @@ $blocks-button__height: 56px;
 	}
 }
 
-.wp-gs .wp-block-button__link:not(.has-background) {
```
See [comment](https://github.com/WordPress/gutenberg/pull/20530/files?file-filters%5B%5D=.js&file-filters%5B%5D=.json&file-filters%5B%5D=.md&file-filters%5B%5D=.scss&hide-deleted-files=true#r400832037).

nosolosw commented on packages/block-library/src/heading/style.scss:
```diff
@@ -7,7 +7,8 @@
 		background-color: var(--wp--color--background);
 		color: var(--wp--color--text);
-		line-height: var(--wp--typography--line-height);
+		line-height: var(--wp--typography--line-height, var(--wp--typography--line-height-heading));
```
Note that, the line-height hook that is shown in the block inspector is using `--wp--typography--line-height`, which is added as an inline style. If not present, it'll fall back to the global variable. This may need iteration in a subsequent PR, we may want to consolidate everything to general variables ([potential direction](https://href.li/?https://github.com/WordPress/gutenberg/pull/20290), although needs discussion).

nosolosw commented on packages/block-library/src/heading/style.scss:
```diff
@@ -21,3 +22,27 @@ h6 {
 	}
 }
+
+:root h1 {
```
In the [global styles UI](https://github.com/WordPress/gutenberg/issues/21030), we aim to expose base and scale, and not font size directly. At the block UI level, we're instead inlining font-size values directly (see [related PR](https://github.com/WordPress/gutenberg/pull/21153) to make font size an implicit attribute of the block). This may need some consolidation in further PRs.

We also may do with some pre-generation of variables or/and with some improvements to the theme.json, as @jorgefilipecosta ([suggested](https://github.com/WordPress/gutenberg/pull/20530#discussion_r391767806)). I think we may want to explore that separately.

kjellr replied:
Can you share a little clarification on this? Currently the styles here auto-generate heading sizes, and do so with the `:root h1 {}` specificity, which overrides all of the usual theme-provided `h1 {}` heading styles. Is that the intended approach? I'm a little unclear how themes are supposed to provide heading sizes at the moment.

jasmussen replied:
I'm so incredibly excited about these features to land, and some of my favorite people are working on this! 

However, and especially seeing the GIF above, the "base font" and "base scale" is the only aspect that feels unintuitive to me. I understand that the idea is to provide a harmonious typescale, and to provide a single control to affect all headings in one go. 

But my first question is "how do I adjust only the H2?"

Tying this to a type-scale only feels like it might end up frustrating users. What if I want my H2 to be almost exactly as big as my H1, and my H3 and H4 to be much smaller? If I'm reading this interface correctly, that's not possible in a global sense. More often than not the designs I've worked with use heading sizes that don't fit a typescale at all, but are still appropriate for the design in question. Many times they even use different font-weights or even fonts at different levels.

Don't get me wrong, the type scale is impressive, and there are many reasons to use one. But to me the type scale feels like an _option_ to offer, rather than being the default or only interface. 

CC: @karmatosed

nosolosw replied:
There are a few things at play here:

- Design-wise, I don't think I can personally provide any more input than what it's being discussed at https://github.com/WordPress/gutenberg/issues/21030#issuecomment-606675549 The current direction is that users & themes should set base & scale. Blocks & Themes should use those variables to set the font-size.

- There is also the `:root h*` bit, I can comment more on this. We're experimenting with adding some CSS variables to blocks by default (more context in [thread](https://github.com/WordPress/gutenberg/pull/21037#discussion_r397068844)). Some examples of current merged changes are line-height for [heading](https://github.com/WordPress/gutenberg/pull/21215/files#diff-f2f5efb13e0e16e61312aa9b0c028ae1R1), color & background-color for [paragraph](https://github.com/WordPress/gutenberg/pull/21037/files#diff-6aca82de6dd8ae08de9708ae9ed73ce6R1), [heading](https://github.com/WordPress/gutenberg/pull/21039/files#diff-f2f5efb13e0e16e61312aa9b0c028ae1R6), and [Media/Text](https://github.com/WordPress/gutenberg/pull/21169/files#diff-eb63570fda7a811f34f5ce8f58ed4d37R1). This PR expands on that approach and adds font-weight and font-size by default to paragraph & heading. The vision is that this allows themes to config the CSS variables via the `theme.json` without needing to use them directly unless they explicitly want to expand what core does or do things differently.

karmatosed replied:
Lots of great conversation here. I wanted to pop in and first say nothing here is set. I think what was mocked up was a good first bet, let's maybe see where this needs iterating.

There are a few options:

- Just a scale: this has limited flexibility, however it could use the body text size to build from.
- Set a base: this feels a more flexible approach but it also has complications.

I can totally see maybe this is theme optional, maybe it all is for base font and type scale. I however, do see that a user could find this really magical being able to go up and down.

Regarding individual headings, this is where I agree there needs to be some considered refinement. We are considering elements in groups, over individually. One option which is debatable if v1 or beyond (I could go either way), could be to have an option for more granular control 'reveal'. A simple then list of each heading could do that for headings. I'm totally into exploring around this.

oandregal committed:
Fix rebase odditiy

nosolosw commented on packages/edit-site/src/components/sidebar/style.scss:
```diff
@@ -15,4 +15,20 @@
 	.block-editor-block-inspector__card {
 		margin: 0;
 	}
+
```
This is so the block styles don't leak into the sidebars (block inspector, global styles).

mtias replied:
Can we add these as comments?

nosolosw replied:
done at https://github.com/WordPress/gutenberg/pull/20530/commits/937578a5ada1fd9c05ad851d5ea27e3686bef66e

nosolosw commented on packages/edit-site/src/style.scss:
```diff
@@ -25,6 +25,10 @@ body.toplevel_page_gutenberg-edit-site {
 	@include reset;
 }
 
+.components-modal__frame h1 {
```
This is so the block styles don't leak into the saving modal (tried with TwentyTwenty that uses a distinct background-color).

nosolosw replied:
Added comment.

nosolosw commented on packages/global-styles/src/renderer.js:
```diff
@@ -0,0 +1,63 @@
+	return toReturn;
+}
+
+function compileStyles( styles = {} ) {
```
For reference, there's another place where we do this kind of thing https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/hooks/style.js#L30 We should look into consolidating somehow.

oandregal committed:
Render the InspectorControl after GlobalStyles

oandregal committed:
Make strings translatable

oandregal committed:
Add comments

jorgefilipecosta commented on packages/block-library/src/heading/style.scss:
```diff
@@ -7,7 +7,8 @@
 		color: var(--wp--color--text);
-		line-height: var(--wp--typography--line-height);
+		line-height: var(--wp--typography--line-height, var(--wp--typography--line-height-heading));
+		font-weight: var(--wp--typography--font-weight-heading);
```
Should we just remove the "-heading" suffix given that we will probably not have block specific variables?

nosolosw replied:
I've commented on this at https://github.com/WordPress/gutenberg/pull/20530#discussion_r400836040 TLDR is that so far we don't have a way to do otherwise. I proposed a [potential direction](https://href.li/?https://github.com/WordPress/gutenberg/pull/20290) but it needs some adjustments before is ready.

This is a discussion that also needs to address what do we do with font-size (used at the custom styles/block level) VS font base/scale (used at the global styles level) I mentioned at https://github.com/WordPress/gutenberg/pull/20530#discussion_r400840203

Those seem conversations we need to have before settling on anything, so I lean towards merging this and then iterate. How does it sound?

jorgefilipecosta commented on packages/edit-site/src/components/sidebar/index.js:
```diff
@@ -4,32 +4,33 @@
 function Sidebar() {
 	return (
-		<>
+		<div className="edit-site-sidebar">
```
Instead of adding a wrapper could we pass the className to ComplementaryArea? This would allow us to avoid rendering the wrapper even if no sidebar is open.
Ideally, we would also avoid calling the class "edit-site-sidebar" and would call it "edit-site-complementary-area".

nosolosw replied:
done at https://github.com/WordPress/gutenberg/pull/20530/commits/f44c84caf52e0ed19a4724a57e7680cd64af3a36

jorgefilipecosta commented on packages/edit-site/src/components/sidebar/index.js:
```diff
@@ -4,32 +4,33 @@
+				icon={ cog }
 			>
-				<p>Global Styles area</p>
+				<InspectorSlot bubblesVirtually />
```
I think now we have bubblesVirtually by default.

jorgefilipecosta commented on packages/edit-site/src/components/sidebar/style.scss:
```diff
@@ -15,4 +15,22 @@
 		margin: 0;
 	}
+
+	// There are to prevent the blocks styles
```
Would it make sense to make these styles part of the interface package?

nosolosw replied:
Done at https://github.com/WordPress/gutenberg/pull/20530/files?file-filters%5B%5D=.scss#diff-ea6e8e5d17ac512ea81dc2085c71bf43R53 This was... tricky :sweat_smile: 

I tested with the document & block sidebars (edit-post) and global & block sidebars (edit-site). Everything seems to be fine, at least I don't find any gross leak now.

jorgefilipecosta reviewed (commented):
Hi @nosolosw when I execute the editor on this branch I see an empty global styles sidebar. Am I doing something wrong?

nosolosw commented:
> When I execute the editor on this branch I see an empty global styles sidebar. Am I doing something wrong?

@jorgefilipecosta Ah, I see that may happen if you use a theme that hasn't declared support via `experimental-theme.json`.  Perhaps we should render a "Your theme doesn't support global styles" kind of message? I don't know whether the preferred behavior, in this case, would be that or do not render the global styles sidebar.

See testing instructions in the issue description but TLDR is I recommend using the theme at https://github.com/WordPress/theme-experiments/pull/22

mtias commented:
> I don't know whether the preferred behavior, in this case, would be that or do not render the global styles sidebar.

Not rendering it all.

jorgefilipecosta commented:
> Ah, I see that may happen if you use a theme that hasn't declared support via experimental-theme.json

Given that now we don't have wp-gs class, and blocks are using CSS variables unconditionally does it still make sense to have this sidebar being opt-in?
Maybe we should render it all the time and then have a way to disable it, as we allow to disable colors, gradients, etc.

oandregal committed:
Pass class to complementaryareas

oandregal committed:
Remove check about edit-site

jorgefilipecosta commented:
Hi @nosolosw,
When I update the variables the changes are not reflected:
![Mar-31-2020 16-37-11](https://user-images.githubusercontent.com/11271197/78046472-18a12700-736f-11ea-8376-7c4d92cfb131.gif)

oandregal committed:
Do not render GlobalStyles if theme has no support

oandregal committed:
Fix rebase oddity

oandregal committed:
Remove unnecesary code

The block's CSS is already setting line-height
and takes precedence from wp-admin styles,
so we no longer need this override.

oandregal committed:
Remove unnecesary color override

The block's CSS is already taken precedence from wp-admin styles,
so we no longer need this.

nosolosw commented on packages/editor/src/editor-styles.scss:
```diff
@@ -42,17 +42,6 @@ h6 {
 	font-size: 0.8em;
 }
 
-// Adjust line spacing for larger headings.
```
:wave: @jasmussen Want to double-check why these styles were useful for? Is it the case that this are here to prevent wp-admin styles from overriding editor styles?

If so, we may be able to get rid of some of them: one of the [changes](https://github.com/WordPress/gutenberg/pull/21215/files#diff-f2f5efb13e0e16e61312aa9b0c028ae1) introduced in the line-height control PR recently was to add a line-height property by default to the headings block. Same for color below in [this PR](https://github.com/WordPress/gutenberg/pull/21215/files#diff-f2f5efb13e0e16e61312aa9b0c028ae1).

jasmussen replied:
Thanks for the ping. Please get rid of as much of this as you like, so as to make sense of the global style system. While I can't 100% recall exactly why I added those spacings in the first place, a very good bet is that it had to do with inherited CSS bleed from wp-admin.

oandregal committed:
Move font-weight resets to interface package

oandregal committed:
Move font-size to interface package

oandregal committed:
line-height reset: adject and move to interface package

jorgefilipecosta commented on packages/edit-site/src/components/sidebar/index.js:
```diff
@@ -4,31 +4,57 @@
+					</ComplementaryArea>
+				</>
+			) : (
+				<ComplementaryArea
```
I guess we don't need to repeat the edit-site/block-inspector area code can we unconditionally render it and then only render "edit-site/global-styles" are if hasGlobalStylesSupport is true?

nosolosw replied:
I hoped so! Unfortunately, I couldn't make it work due to a weird interaction of block settings & the complementary area slot. This is what happens when we do that:

- block settings for global styles are initially false => the slot renders the inspector control
- block settings for global styles are updated to true => the global styles area is **appended** to the existing areas, so it ends up in the last position.

jorgefilipecosta replied:
Ah got it, this happens because the order of the fills is not relevant to the order in which they are rendered.
In that case, I suggest, defining a variable with the inspector component to avoid reporting its declaration:
```
const blockInspector = ( <ComplementaryArea
				scope="core/edit-site"
				complementaryAreaIdentifier="edit-site/block-inspector"
				title={ __( 'Block Inspector' ) }
				icon={ cog }
				className="edit-site-complementary-area"
			>
				<InspectorSlot bubblesVirtually />
			</ComplementaryArea>
)
```

oandregal committed:
Fix h2 line-height

oandregal committed:
Fix color for paragraphs

oandregal committed:
Add fix for background-color

oandregal committed:
Moved to complementary area

jorgefilipecosta commented on packages/edit-site/src/components/sidebar/index.js:
```diff
@@ -4,31 +4,57 @@
 );
 function Sidebar() {
+	const hasGlobalStylesSupport = useSelect( ( select ) => {
+		return Object.keys( select( 'core/block-editor' ).getSettings() ).some(
```
Could we do `return !! select( 'core/block-editor' ).getSettings(). __experimentalGlobalStylesUserEntityId;`?

jorgefilipecosta commented on packages/block-library/src/heading/style.scss:
```diff
@@ -21,3 +22,27 @@ h6 {
+}
+
+:root h6 {
+	font-size: calc(0.5px * var(--wp--typography--font-base) * var(--wp--typography--font-scale));
```
I think there is some probability that they will break many teams.
We are only allowing to use global styles if theme has some kind of support, but we are unconditionally adding the styles that use the global styles variables.
I think given that styles are unconditional the global styles should also be. So, at least themes have a way to fix things using theme.json if the result is not expected.

nosolosw replied:
This PR adds font-size and font-weight to existing blocks that have already added CSS to target line-height and/or colors as CSS variables. Is there any reason we should treat font-size or font-weight differently?

jorgefilipecosta replied:
For line height, and colors we use the variables but these variables can be defined by the user. E.g: the user can use the UI to change line-height, and colors.

Here we use the variables and don't allow any control over the variables in most themes.
I think we should uncondiotanily show the global styles UI because we unconditionally add the styles that use the variables.

nosolosw replied:
Ah, I see what you mean, thanks for the clarification :+1:

Here's my thinking:

- I think we can land this PR without user-facing UI controls: UI controls at the block level are nice, but they don't fix the potential issues. Example: for colors, the vars are only used when the users choose a _custom_ color but the CSS is always present so the theme still needs to account for when the user doesn't modify anything. However! I've already started adapting [the UI controls to work with font-size](https://github.com/WordPress/gutenberg/pull/21153), I just don't think that work should block this.

- As to whether to always load the global styles mechanism. Your point is solid: if we already ship block's CSS with variables, we should allow themes to set them. I may be on board with that, let's discuss it separately at https://github.com/WordPress/gutenberg/pull/21346

nosolosw commented:
For everyone following this thread: I'm going to let this branch sit for a few days as it is while I look at other PRs ([font-size attribute](https://github.com/WordPress/gutenberg/pull/21153), [targeted selectors](https://github.com/WordPress/gutenberg/pull/20290)).

jorgefilipecosta commented:
Closing this PR as it seems it was supersede by https://github.com/WordPress/gutenberg/pull/24250.
